### PR TITLE
Mailbox/set update mailbox name

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 1]
```diff
@@ -1600,4 +1601,782 @@ trait MailboxSetMethodContract {
...
+         |					},
+         |					"isSubscribed": true
+         |				}
+         |			}
+         |		}, "c1"],
+         |		["Mailbox/set", {
+         |			"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |			"newState": "000001",
+         |			"updated": {
+         |				"7": {}
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:1805]: this mailboxId needs to be a variable

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 2]
```diff
@@ -1600,4 +1601,782 @@ trait MailboxSetMethodContract {
...
+         |	]
+         |} {
+         |	"sessionState": "75128aab4b1b",
+         |	"methodResponses": [
+         |		["Mailbox/set", {
+         |			"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |			"newState": "000001",
+         |			"created": {
+         |				"C43": {
+         |					"id": "7",
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:1781]: this mailboxId needs to be a variable

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 3]
```diff
@@ -1600,4 +1601,782 @@ trait MailboxSetMethodContract {
...
+         |		}, "c1"],
+         |		["Mailbox/set", {
+         |			"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |			"newState": "000001",
+         |			"updated": {
+         |				"$mailboxId": {}
+         |			}
+         |		}, "c2"]
+         |	]
+         |} {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:1773]: Why is there two objects?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 4]
```diff
@@ -1600,4 +1601,782 @@ trait MailboxSetMethodContract {
...
+         |	"sessionState": "75128aab4b1b",
+         |	"methodResponses": [
+         |		["Mailbox/set", {
+         |			"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |			"newState": "000001",
+         |			"notUpdated": {
+         |				"${mailboxId.serialize()}": {
+         |					"type": "invalidArguments",
+         |					"description": "Invalid change to a system mailbox",
+         |          "properties":["/name"]
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:2326]: indent

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala [hunk 5]
```diff
@@ -124,10 +155,76 @@ class MailboxSetMethod @Inject()(serializer: Serializer,
...
+      // No updated properties passed. Noop.
+      .getOrElse(SMono.just[UpdateResult](UpdateSuccess(mailboxId)))
+  }
+
+  private def computeMailboxPath(mailbox: MessageManager, nameUpdate: NameUpdate, mailboxSession: MailboxSession): MailboxPath = {
+    val originalPath: MailboxPath = mailbox.getMailboxPath
+    val maybeParentPath: Option[MailboxPath] = originalPath.getHierarchyLevels(mailboxSession.getPathDelimiter)
+      .asScala
+      .reverse
+      .drop(1)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala:222]: TODO add a test renaming a child mailbox

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala:222]: I also added the test concerning renaming parent mailboxes

**Comment by chibenwa** [comment]
TODO add ticket number https://issues.apache.org/jira/browse/JAMES-3359

**Comment by chibenwa** [comment]
test this please

failed in deployment tests

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala [hunk 6]
```diff
@@ -225,9 +225,24 @@ class Serializer @Inject() (mailboxIdFactory: MailboxId.Factory) {
...
+      mapWithStringKey
+        .foldLeft[Either[JsError, Map[UnparsedMailboxId, MailboxPatchObject]]](scala.util.Right[JsError, Map[UnparsedMailboxId, MailboxPatchObject]](Map.empty))((acc: Either[JsError, Map[UnparsedMailboxId, MailboxPatchObject]], keyValue) => {
+          acc match {
+            case error@Left(_) => error
+            case scala.util.Right(validatedAcc) =>
+              val refinedKey: Either[String, UnparsedMailboxId] = refineV(keyValue._1)
+              refinedKey match {
+                case Left(error) => Left(JsError(error))
+                case scala.util.Right(unparsedMailboxId) => scala.util.Right(validatedAcc + (unparsedMailboxId -> keyValue._2))
+              }
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala:239]: Maybe would be clearer to extract the refine into a method with the `UnparsedMailboxId` type?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 7]
```diff
@@ -1772,6 +1773,48 @@ trait MailboxSetMethodContract {
          |}""".stripMargin)
   }
 
+  @Disabled("TODO implementmailbox/set creation subscription support and enable upon rebase")
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:1776]: Is that for an other ticket/PR?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 8]
```diff
@@ -2033,6 +2033,108 @@ trait MailboxSetMethodContract {
...
+         |	"sessionState": "75128aab4b1b",
+         |	"methodResponses": [
+         |		["Mailbox/set", {
+         |			"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |			"newState": "000001",
+         |			"notUpdated": {
+         |				"${mailboxId.serialize()}": {
+         |					"type": "invalidArguments",
+         |					"description": "Invalid change to a system mailbox",
+         |          "properties":{"value":["/name"]}
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala]: indent

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala]: Outdated... my bad

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala (hunk 6 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala:239]: UnparsedMailboxId is a type and not a class thus can't receive methods.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala (hunk 7 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:1776]: See https://github.com/linagora/james-project/pull/3661

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Rebased to solve a conflict with master.

Only the last fixup was added. 

I leveraged this opportunity to the ticket number in commit messages, addressing https://github.com/linagora/james-project/pull/3665#issuecomment-674783018

**Comment by chibenwa** [comment]
```
[8e762d6686a562660874d85cfd8fbfa814e56a19] Failed scenarios:
[8e762d6686a562660874d85cfd8fbfa814e56a19] cucumber/sharing/MoveMailboxAndSharing.feature:72 # A sharee should be able to retrieve a mailbox after sharer moved it into another mailbox
[8e762d6686a562660874d85cfd8fbfa814e56a19] 
[8e762d6686a562660874d85cfd8fbfa814e56a19] 2 Scenarios (1 failed, 1 passed)
[8e762d6686a562660874d85cfd8fbfa814e56a19] 20 Steps (1 failed, 19 passed)
[8e762d6686a562660874d85cfd8fbfa814e56a19] 5m46.601s
[8e762d6686a562660874d85cfd8fbfa814e56a19] 
[8e762d6686a562660874d85cfd8fbfa814e56a19] java.lang.AssertionError: 
[8e762d6686a562660874d85cfd8fbfa814e56a19] Expecting:
[8e762d6686a562660874d85cfd8fbfa814e56a19]   <[null]>
[8e762d6686a562660874d85cfd8fbfa814e56a19] to contain only:
[8e762d6686a562660874d85cfd8fbfa814e56a19]   <["fa5ff280-e204-11ea-8672-6b9c64e52fee"]>
[8e762d6686a562660874d85cfd8fbfa814e56a19] elements not found:
[8e762d6686a562660874d85cfd8fbfa814e56a19]   <["fa5ff280-e204-11ea-8672-6b9c64e52fee"]>
[8e762d6686a562660874d85cfd8fbfa814e56a19] and elements not expected:
[8e762d6686a562660874d85cfd8fbfa814e56a19]   <[null]>
[8e762d6686a562660874d85cfd8fbfa814e56a19] 
[8e762d6686a562660874d85cfd8fbfa814e56a19] 	at org.apache.james.jmap.draft.methods.integration.cucumber.GetMailboxesMethodStepdefs.assertMailboxesNames(GetMailboxesMethodStepdefs.java:95)
[8e762d6686a562660874d85cfd8fbfa814e56a19] 	at ✽.the mailboxes should contain "sharedChild" in "Delegated" namespace, with parent mailbox "notShared" of user "alice@domain.tld"(cucumber/sharing/MoveMailboxAndSharing.feature:76)
```

**Comment by chibenwa** [comment]
Weird.

Passes locally. Changes are contained in new RFC implementation thus should not impact draft.

test this please

**Comment by chibenwa** [comment]
Deployment tests. Can be considered :green_apple: .

test this please

**Review by LanKhuat — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Review by rouazana — COMMENTED** [review]
I don't see any test case around mailbox separator, is it expected?

**Comment by chibenwa** [comment]
Added kn an other pr

**Pull request closed without merging**